Improve intro to machine docs

londoncalling opened the pull request:
Signed-off-by: Victoria Bialas <victoria.bialas@docker.com>

londoncalling commented:
@moxiegirl this is still in work per the "WIP: prefix", but wanted to push current changes, please do not merge yet :-)

londoncalling commented:
@moxiegirl maybe we can take a look at the output of "make docs" on this? At this point, the Machine "overview" isn't even in the menus at all now , and the build is showing a link error ... at some point, the menus being in flux  might make it difficult for me to check my work with builds, as I continue updating these Machine docs ...

londoncalling commented:
@moxiegirl @nathanleclaire @dgageot @bfirsh @jeanlaurent @dmp42 

Please review these updated Machine docs , and approve merge if ok. 

Since there quite a few changes, I've published this draft in a make docs container on Digital Ocean (a nice test of that `docker-machine` driver) so that you can read them as a cohesive docset: http://104.131.118.86:8000/machine/overview/ 

(You will see nav menus below the Machine docs without page ID codes instead of topic names--please ignore, this is work-in-progress docs infrastructure.)

Key changes are:
*  a rewrite of "overview" 
* re-org of "getting started with a local VM"
* topic re-organization and copyedits throughout for clarity

The last thing I'd like to do, per @moxiegirl suggestion is remove the Swarm topics at the end of the "cloud provider" topic (making sure this is covered in Swarm docs)

Let me know what you'd like changed or if this is good-to-go. Thanks! Vicky

moxiegirl commented:
@londoncalling etc.  The new provisioning documentation for Swarm:

https://github.com/docker/swarm/issues/1719

chanwit commented on docs/concepts.md:
```diff
@@ -0,0 +1,94 @@
+
+Since Docker runs on Linux, each VM that Docker Machine provisions relies on a
+base operating system. For convenience, there are default base operating
+systems. For the Oracle Virtual Box driver, this base operating system is the
```
the Oracle VirtualBox driver

chanwit commented on docs/concepts.md:
```diff
@@ -0,0 +1,94 @@
+Before you can run a `docker` command on a machine, you configure your
+command-line to point to that machine. The `docker-machine env <machine-name>`
+subcommand outputs the configuration command you should use. When you run a
+container on the Docker host, the container's ports map to ports on the VM.
```
There is also a `generic` driver, which allows provisioning a physical machine. So this would change to:
"the container's ports map to ports on the machine".

chanwit commented on docs/concepts.md:
```diff
@@ -0,0 +1,94 @@
+host on a VM and you use the Docker Engine client as needed to build images and
+create containers on the host.
+
+## Drivers for creating virtual machines
```
I think it should be "Drivers for creating machines".

chanwit commented on docs/concepts.md:
```diff
@@ -0,0 +1,94 @@
+
+# Understand Machine concepts and get help
+
+Docker Machine allows you to provision Docker on virtual machines that reside
```
We can also provision a bare metal, so this sentence maybe incorrect.

bfirsh replied:
Yeah, maybe "Docker Machine allows you to provision Docker Engine on either your local system, on cloud providers, or on bare metal servers ..."

bfirsh replied:
But tbh, vast majority of people will be creating virtual machines. Perhaps the fact that it can provision existing servers could be a side note.

dmp42 replied:
Machine lets you provision on a variety of environments. Starting with such a statement (eg: limiting to virtual machines) is misleading.

dmp42 replied:
eg: provisioning on, say, "digital ocean" shouldn't make assumption about the underlying architecture.

chanwit commented on docs/concepts.md:
```diff
@@ -0,0 +1,94 @@
+## Drivers for creating virtual machines
+
+To create a virtual machine, you supply Docker Machine with the name of the
+driver you want use. The driver represents the virtual environment. For example,
```
> The driver represents the virtual environment. For example, on a local Linux, Mac, or Windows system the driver is typically Oracle VirtualBox.

This statement is also not correct for `generic` and SoftLayer's drivers.

londoncalling replied:
Thanks Chamwit, I'll get these corrections in. Much of this is legacy docs
but great to catch now !

On Sunday, January 31, 2016, Chanwit Kaewkasi <notifications@github.com>
wrote:

> In docs/concepts.md
> <https://github.com/docker/machine/pull/2957#discussion_r51366143>:
>
> > +weight=3
> > ++++
> > +<![end-metadata]-->
> > +
> > +
> > +# Understand Machine concepts and get help
> > +
> > +Docker Machine allows you to provision Docker on virtual machines that reside
> > +either on your local system or on a cloud provider. Docker Machine creates a
> > +host on a VM and you use the Docker Engine client as needed to build images and
> > +create containers on the host.
> > +
> > +## Drivers for creating virtual machines
> > +
> > +To create a virtual machine, you supply Docker Machine with the name of the
> > +driver you want use. The driver represents the virtual environment. For example,
>
> The driver represents the virtual environment. For example, on a local
> Linux, Mac, or Windows system the driver is typically Oracle VirtualBox.
>
> This statement is also not correct for generic and SoftLayer's drivers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/machine/pull/2957/files#r51366143>.
>

londoncalling replied:
Hi Chanwit,
I believe I incorporated all your comments. Please take a look at the
updated Concepts topic on GitHub for this PR,
<https://github.com/docker/machine/pull/2957#issuecomment-177574034> and/or
at Machine concepts and help <http://104.131.118.86:8000/machine/concepts/>
on the draft docs site.

Thanks for the comments--all good catches. --Vicky

On Sun, Jan 31, 2016 at 11:55 AM, Chanwit Kaewkasi <notifications@github.com
> wrote:

> In docs/concepts.md
> <https://github.com/docker/machine/pull/2957#discussion_r51366143>:
>
> > +weight=3
> > ++++
> > +<![end-metadata]-->
> > +
> > +
> > +# Understand Machine concepts and get help
> > +
> > +Docker Machine allows you to provision Docker on virtual machines that reside
> > +either on your local system or on a cloud provider. Docker Machine creates a
> > +host on a VM and you use the Docker Engine client as needed to build images and
> > +create containers on the host.
> > +
> > +## Drivers for creating virtual machines
> > +
> > +To create a virtual machine, you supply Docker Machine with the name of the
> > +driver you want use. The driver represents the virtual environment. For example,
>
> The driver represents the virtual environment. For example, on a local
> Linux, Mac, or Windows system the driver is typically Oracle VirtualBox.
>
> This statement is also not correct for generic and SoftLayer's drivers.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/machine/pull/2957/files#r51366143>.
>

londoncalling commented:
Thanks @moxiegirl

On Sunday, January 31, 2016, moxiegirl <notifications@github.com> wrote:

> @londoncalling <https://github.com/londoncalling> etc. The new
> provisioning documentation for Swarm:
>
> docker/swarm#1719 <https://github.com/docker/swarm/pull/1719>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/docker/machine/pull/2957#issuecomment-177574034>.
>

londoncalling commented:
I've updated these docs as follows.

* Corrected some details in Concepts (http://104.131.118.86:8000/machine/concepts/) per comments from Chanwit

* Replaced info on Swarm clusters in Cloud topic (http://104.131.118.86:8000/machine/get-started-cloud/)  with a reference to the Swarm docs on using Machine to provision Swarm clusters per @moxiegirl. Links won't work yet.  (@moxiegirl, please let me know if I anticipated these links out to Swarm topics correctly - thanks!)

Please review here or at http://104.131.118.86:8000/machine/overview/. Reviewers, please let me know if this pull request can be merged now.

bfirsh commented on docs/concepts.md:
```diff
@@ -0,0 +1,95 @@
+## Drivers for creating machines
+
+To create a virtual machine, you supply Docker Machine with the name of the
+driver you want use. The driver represents the virtual environment. For example,
```
I'm not sure this makes sense: "The driver represents the virtual environment."

Maybe something like this? "The driver determines where the virtual machine is created."

londoncalling commented:
@bfirsh brilliant, Ben, thanks. I had re-worded these phrases similarly, but your wording is better. Updates on "[concepts](http://192.168.99.100:8000/machine/concepts/)" are in per your comments.

londoncalling commented:
@moxiegirl I've added in comments from you, @bfirsh and Chanwit, and also now incorporated the illustrated step-by-step example of cloud provisioning w/Digital Ocean example (barring content re-use, for now this is a copy, but it works well here). 

If there are no more comments from @nathanleclaire or rest of team, I believe this is ready to be merged.

The thing to keep track of here is to make sure links to new/moved topics work (e.g., links to the Swarm docs, and to any of my cloud install topics). Let me know how I can help check for those at release time.

nathanleclaire commented on docs/concepts.md:
```diff
@@ -0,0 +1,94 @@
+Since Docker runs on Linux, each VM that Docker Machine provisions relies on a
+base operating system. For convenience, there are default base operating
+systems. For the Oracle Virtual Box driver, this base operating system is the
+`boot2docker.iso`. For drivers used to connect to cloud providers, the base
```
I would expect it to say "this base operation system is boot2docker" with a link to https://github.com/boot2docker/boot2docker

londoncalling commented:
@nathanleclaire thanks for the comment, I updated [Concepts](http://192.168.99.100:8000/machine/concepts/) per that boot2docker reference.

Just a friendly reminder, everyone has been commenting only on  Concepts, but several files have changed markedly. If all LGTYou, great! Just let me know. 

Files with quite a bit of new content or reorganization are:
*  [overview.md](http://192.168.99.100:8000/machine/overview/)
* [get-started.md](http://192.168.99.100:8000/machine/get-started/)
* [get-started-cloud.md](http://192.168.99.100:8000/machine/get-started-cloud/) (which will change again if I can get the AWS example working, then I'll just point to this version of the Digital Ocean example in install docs and use AWS for Machine docs)

dmp42 commented on docs/concepts.md:
```diff
@@ -0,0 +1,91 @@
+## Configuring CLI environment variables for a Docker host
+
+Before you can run a `docker` command on a machine, you configure your
+command-line to point to that machine. The `docker-machine env <machine-name>`
```
s/you configure/you need to configure/ ?

dmp42 commented on docs/concepts.md:
```diff
@@ -0,0 +1,91 @@
+Before you can run a `docker` command on a machine, you configure your
+command-line to point to that machine. The `docker-machine env <machine-name>`
+subcommand outputs the configuration command you should use. When you run a
+container on the Docker host, the container's ports map to ports on the machine.
```
This sounds weird. Containers ports are not exposed unless you do explicitly request them to be. Also wondering if this is adding value - if the docker daemon is reachable via a certain ip address, then exposed container ports are necessarily available on that ip (eg: I guess my point is that this is not machine specific).

dmp42 commented on docs/concepts.md:
```diff
@@ -0,0 +1,91 @@
+or firewall issues.  There are also reasons from the other end of the chain:
+your cloud provider or the network in between.
+
+To help `docker-machine` be as stable as possible, we added a monitoring of
```
s/a monitoring of crashes whenever you try to `create` or `upgrade` a host/crash reporting support on `create` and `upgrade` operations/ ?

dmp42 commented on docs/get-started-cloud.md:
```diff
@@ -9,176 +9,252 @@ weight=2
-
-From this point, the remote host behaves much like the local host we created in the last section. If we look at `docker-machine ls`, we'll see it is now the
-"active" host, indicated by an asterisk (`*`) in that column:
+  1. Go to the Digital Ocean administrator console and click **API** in the header.
```
It feels weird that we provide such detailed explanations about how to setup an account on a given cloud provider.

First, it gives the feeling that we are endorsing its use over the others. Second, this is very basic operations, that should be/are documented by the cloud provider himself.

Could we instead point to their documentation? (that would also guarantee it's up to date instead of adding extra maintenance on us).

Let me know what you think.

londoncalling replied:
@dmp42 , @moxiegirl we had a request from @shykes in terms of Cloud installs, not to point to third party docs. As a result, I created a PR for this at https://github.com/docker/docker/pull/18627, and updated the Cloud install documentation based on that request). Maybe we should touch base on this whole strategy to make sure we are on the same page.

dmp42 replied:
@londoncalling do you have a link to that conversation with @shykes so to better understand the context and figure out a good solution here?
Thanks.

londoncalling replied:
@dmp42 , I don't have the text of the convo, but @moxiegirl does ... it started between them.

dmp42 commented on docs/get-started-cloud.md:
```diff
@@ -9,176 +9,252 @@ weight=2
+    This is the personal access token you'll use in the next step to create your cloud server.
 
-### 3rd-party driver plugins
+### Step 3. Start Docker Machine
```
I'm confused about this Step 3 section. Why would I need a local VM to be started to provision on digital ocean?

londoncalling replied:
@dmp42 , good point .. this section needs revision. I do want to have them run some `docker` commands to make sure that both `docker-machine` and `docker` are working, but the examples in Step 3 are confusing. The only other thing I'd mention here, is that if you installed with Toolbox, you will have a default machine, so I might still be tempted to show it here. But, yes, this section needs revision ... coming up.

dmp42 commented on docs/get-started-cloud.md:
```diff
@@ -9,176 +9,252 @@ weight=2
+        docker-sandbox   *        digitalocean   Running   tcp://45.55.222.72:2376     
 
-### Swarm master
+6. Log in to the Droplet with the `docker-machine ssh` command.
```
You don't need to do that, and this should be discouraged, since the purpose of machine is to (some extent) remove the need from ssh-ing into remote machines.

To check that the docker engine is working, just stay where you are (your local environment) and use docker itself.

londoncalling replied:
@dmp42 , agreed that the benefit of `docker-machine` is that you typically do not have to `ssh` in at all. However, in cases where you might need or want to (troubleshooting, navigating logs, curiosity / investigating your instance) .. I'd like to showcase that `docker-machine` makes this much easier: 

``` 
docker-machine ssh <machine>
```

vs.

```
ssh -i "my-key-pair.pem" ubuntu@xx.xxx.xxx.xxx
```
... And it's just cool that you can do this with Machine.

your thoughts ...?

dmp42 replied:
@londoncalling agreed.
My point is that at this point of the explanation, what makes the most sense is to use docker from where you sit to check the remote is working (instead of ssh-ing).

eg:

 * machine create -driver digitalfoo superfoo
 * eval $(machine superfoo)
 * docker pull bar
 * -> it's a win

Adding an ssh step in between does muddy the workflow a bit, and though useful in itself, IMHO belong elsewhere.

Hope that clarify.

londoncalling replied:
@dmp42 Yes, that makes sense ... I'll remove this here, and add it in somewhere else, so as not to confuse the flow. Good idea, thanks.

dmp42 commented on docs/get-started-cloud.md:
```diff
@@ -9,176 +9,252 @@ weight=2
+
+## Adding a host without a driver
+
+You can add a host to Docker which only has a URL and no driver. Therefore it can be used an alias for an existing host so you don’t have to type out the URL every time you run a Docker command.
```
used as an alias?

londoncalling commented:
@dmp42 @nathanleclaire @moxiegirl Most recent changes here are in response to Olivier's comments, which I believe are all addressed (and greatly improved the topic!) Please let me know if this looks good, and if so give me an LGTM for merge. (We're hammering for release, so soon would be good! :-)

Thanks, Vicky

londoncalling commented:
@moxiegirl pushed latest updates, should be good-to-go but need LGTM from @dmp42 or @nathanleclaire

nathanleclaire commented:
Looking over it now

nathanleclaire commented:
Content LGTM, looks like Travis is still building.

moxiegirl commented:
LGTM
